### PR TITLE
Make parser more robust

### PR DESCRIPTION
**Pull request by mcastorina**
Merging exception handling and tokenizer

### CODE REVIEW

File: slack/dice.py [hunk 1]
```diff
@@ -1,28 +1,44 @@
...
+    except ZeroDivisionError as e:
+        msg = 'Division by zero (%s/%s)' % (str(a), str(b))
+        error = e
+    except OverflowError as e:
+        msg = 'Overflow error (%s, %s)' % (str(a), str(b))
+        error = e
+    except Exception as e:
+        msg = 'Unknown error: %s' % e
+        error = e
+    raise DiceRollParseError(msg).with_traceback(sys.exc_info()[2]) from error
```
Comment (kilogram) [inline slack/dice.py]: https://docs.python.org/3/library/sys.html#sys.exc_info:
> The information returned is specific both to the current thread and to the current stack frame. If the current stack frame is not handling an exception, the information is taken from the calling stack frame, or its caller, and so on until a stack frame is found that is handling an exception.

As it is, I think in this case exc_info will fail, right?

Comment (kilogram) [inline slack/dice.py]: I don't think with_traceback is necessary in this case.

```
def a(limit=5):
    try:
        if limit <= 0:
            raise ValueError("invalid limit")
        return a(limit=limit-1)
    except ValueError as e:
        err = e
    except IndexError as e:
        err = e
    raise NotImplementedError from e

a()
```

gives a full stacktrace.

File: slack/dice.py [hunk 2]
```diff
@@ -1,28 +1,44 @@
 import re
+import sys
 import random as rand
 from io import StringIO
 from functools import reduce
 from operator import mul
 MAX_ITERATIONS = 5000
```
Comment (kilogram) [inline slack/dice.py]: Can we make this a setting as in https://github.com/kilogram/autogm/blob/05bbfc76c4dfaf1614a0c6df143a61dd0a76cfbb/autogm/settings.py#L89

File: slack/dice.py [hunk 3]
```diff
@@ -1,28 +1,44 @@
 import re
+import sys
```
Comment (kilogram) [inline slack/dice.py]: cleanup if not needed

File: slack/dice.py (hunk 2 above)
Comment (kilogram) [inline slack/dice.py]: Actually on second  thought I'd rather keep this free of all django stuff. Maybe an optional argument to the parser, and the caller reads the limit from settings? A problem for another time.

Comment (mcastorina) [inline slack/dice.py]: By "this" you mean MAX_ITERATIONS?

File: slack/dice.py (hunk 1 above)
Comment (mcastorina) [inline slack/dice.py]: Ah, it's removed now

**Pull request merged**